### DISCUSSION TITLE
Updating version of OpenZFS-on-Windows

### DISCUSSION THREAD

**Discussion by sskras** [openzfsonwindows/openzfs#287] [General] [answered]
This probably belongs more to a QnA / Discussions section (which I miss in this repo).

I have installed `2.1.6rc3`:
```
saukrs@DESKTOP-O7JE7JE MSYS ~
$ sudo zfs version
zfswin-2.1.6rc3
zfs-kmod-zfswin-2.1.6rc3
```

What is the reliable way to update this thing to Eg. newest `2.2.0rc4` ?

Should I just download [`OpenZFSOnWindows-debug-2.2.0rc4.exe`](https://github.com/openzfsonwindows/openzfs/releases/download/zfswin-2.2.0rc4/OpenZFSOnWindows-debug-2.2.0rc4.exe) and run the installer?

**Comment by lundman** [comment]
I believe that should work, if it doesn't, I'll have to fix something :)

**Comment by sskras** [comment]
OK, I went without uninstalling the old version before that.

It asked me if would like to reboot just now.  Is necessary to reboot the OS to activate the new driver?

**Comment by lundman** [comment]
Yeah maybe. Not sure, you can unload it if you export everything first, but to load a new one, Windows seems to want to reboot. Might be a trick to it, I am unsure.

**Comment by sskras** [comment]
Thanks for the insights.

I going to keep the issue open for a while.  It would serve as a reminder to do some RnD about updates without reboot (I have no pools imported mounted at the moment).

But if you wish, feel free to close it.

**Comment by sskras** [comment]
Maybe this should be converted into a discussion? (To keep the list cleaner:)
https://docs.github.com/en/discussions/managing-discussions-for-your-community/moderating-discussions#converting-an-issue-to-a-discussion

**Comment by lundman** [comment, marked as answer]
If anyone wants to attempt this, an outline would be:

***

level 1:
Update `zfsinstaller.exe`, which is a separate minitool in this repo. It is not connected to ZFS at all, and more of a pure Windows tool. C++ ?

https://github.com/openzfsonwindows/openzfs/tree/windows/cmd/os/windows/zfsinstaller

* Execute `zpool.exe export -a`. Either by literally executing `zpool.exe`, or if preferred, link with libzpool.a and duplicate with it does internally:
```
	return (for_each_pool(argc, argv, B_TRUE, NULL,
		    ZFS_TYPE_POOL, B_FALSE, zpool_export_one, &cb));
```

If `zpool.exe export -a` fails, stop. print error.

Then let `zfsinstaller.exe` continue, it will issue UNLOAD to `openzfs.sys` and unload. Exit. There may need to be some tweaking with `DefaultUninstall 128` in `zfsinstaller.cpp`. I seem to remember `128` means "requires reboot".

The next invocation of `zfsinstaller install` should install new drivers. 

***
level 2

Do the same as in level 1, but remember the pools imported before `zpool export -a`, so we can re-import them if a new driver is successfully installed. Smooth like butter.

Ditch the cmdline console, use popups, ask "OK to export" and all that jazz.

But some of that can be part of the "I wish we had it" menubar icon UI "we are just about to implement" in similar vein to `ZetaWatch`.

**Comment by derritter88** [comment]
@lundman Thanks for sharing this information.

In general is it possible to do the update without exporting the pools as application would loose connection to the data stored on the pool.
Also it would be usefule to add those information about update process directly at the Github repo README.

**Comment by lundman** [comment]
You would either export the pools, do the upgrade, and in future avoid a reboot.
Or do the upgrade, and reboot to start using the new driver.
I don't see a way to do a live-core update without export/unmounting

**Comment by derritter88** [comment]
Rebooting after the driver update is totally fine for me.
I didn't get it earlier. I thought that you need to export the pool, update, reboot and import the pool again.
